File: ios/Runner/BILMedicalBleBridge.swift
import CoreBluetooth
import Flutter

final class BILMedicalBleBridge: NSObject, CBCentralManagerDelegate, CBPeripheralDelegate {
  static let channelName = "bil.global/medical_ble"
  private let channel: FlutterMethodChannel
  private var central: CBCentralManager!
  private var peripherals: [UUID: CBPeripheral] = [:]
  private var advertisedProfiles: [UUID: Set<String>] = [:]
  private var sessions: [UUID: ReadSession] = [:]
  private var discoveryResult: FlutterResult?
  private var seenPackets = Set<String>()
  private var pendingPairResults: [UUID: FlutterResult] = [:]

  private final class ReadSession {
    let result: FlutterResult
    var packets = [[String: Any]]()
    var expectedCharacteristics = Set<String>()
    var completedCharacteristics = Set<String>()
    var pendingServiceDiscoveries = 0
    var completed = false
    var timeout: DispatchWorkItem?
    init(result: @escaping FlutterResult) { self.result = result }
  }

  static func register(with registrar: FlutterPluginRegistrar) { _ = BILMedicalBleBridge(registrar: registrar) }
  init(registrar: FlutterPluginRegistrar) {
    channel = FlutterMethodChannel(name: Self.channelName, binaryMessenger: registrar.messenger())
    super.init(); central = CBCentralManager(delegate: self, queue: nil); channel.setMethodCallHandler(handle)
  }

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard central.state == .poweredOn else { result(FlutterError(code:"bluetooth_disabled", message:"Bluetooth unavailable", details:nil)); return }
    let args = call.arguments as? [String:Any] ?? [:]
    switch call.method {
    case "requestPermissions":
      // CoreBluetooth presents its system authorization prompt when scanning.
      result(nil)
    case "discover":
      guard discoveryResult == nil else { result(FlutterError(code:"operation_in_progress",message:nil,details:nil)); return }
      discoveryResult = result; central.scanForPeripherals(withServices: supportedServices, options:[CBCentralManagerScanOptionAllowDuplicatesKey:false])
      DispatchQueue.main.asyncAfter(deadline:.now()+3) { self.central.stopScan(); let callback=self.discoveryResult; self.discoveryResult=nil; callback?(self.peripherals.values.map(self.describe)) }
    case "pair":
      guard let p=peripheral(args) else { result(FlutterError(code:"not_found",message:nil,details:nil)); return }
      if p.state == .connected { result(nil); return }
      guard pendingPairResults[p.identifier] == nil else { result(FlutterError(code:"operation_in_progress",message:nil,details:nil)); return }
      pendingPairResults[p.identifier] = result
      central.connect(p)
      DispatchQueue.main.asyncAfter(deadline: .now() + 20) { [weak self] in
        guard let callback = self?.pendingPairResults.removeValue(forKey: p.identifier) else { return }
        callback(FlutterError(code:"pairing_timeout",message:nil,details:nil))
        self?.central.cancelPeripheralConnection(p)
      }
    case "disconnect", "cancel":
      guard let p=peripheral(args) else { result(nil); return }; finish(p.identifier, error:nil); central.cancelPeripheralConnection(p); result(nil)
    case "deviceStatus":
      guard let p=peripheral(args) else { result(["connected":false,"batteryVerified":false]); return }
      result(["connected":p.state == .connected,"batteryPercent":NSNull(),"batteryVerified":false])
    case "forget":
      guard let p=peripheral(args) else { result(nil); return }
      finish(p.identifier,error:nil); central.cancelPeripheralConnection(p); peripherals.removeValue(forKey:p.identifier); advertisedProfiles.removeValue(forKey:p.identifier)
      result(["forgottenLocally":true,"systemUnpairRequired":false])
    case "readMeasurements":
      guard let p=peripheral(args) else { result(FlutterError(code:"not_found",message:nil,details:nil)); return }
      if sessions[p.identifier] != nil { result(FlutterError(code:"operation_in_progress",message:nil,details:nil)); return }
      let session=ReadSession(result:result); sessions[p.identifier]=session; p.delegate=self
      let timeout=DispatchWorkItem { [weak self] in self?.finish(p.identifier,error:FlutterError(code:"gatt_timeout",message:nil,details:nil)) }; session.timeout=timeout
      DispatchQueue.main.asyncAfter(deadline:.now()+20,execute:timeout)
      if p.state == .connected { p.discoverServices(supportedServices) } else { central.connect(p) }
    default: result(FlutterMethodNotImplemented)
    }
  }

  private func peripheral(_ args:[String:Any])->CBPeripheral? { guard let id=args["peripheralId"] as? String, let uuid=UUID(uuidString:id) else{return nil}; return peripherals[uuid] }
  private var supportedServices:[CBUUID]{["181D","181B","180D"].map(CBUUID.init(string:))}
  private var supportedCharacteristics:Set<String>{["2A9D","2A9C","2A37"]}
  func centralManagerDidUpdateState(_ central:CBCentralManager) {}
  func centralManager(_ central:CBCentralManager,didDiscover peripheral:CBPeripheral,advertisementData:[String:Any],rssi RSSI:NSNumber){
    let supported = Set(supportedServices.map { $0.uuidString })
    let advertised = Set((advertisementData[CBAdvertisementDataServiceUUIDsKey] as? [CBUUID] ?? []).map { $0.uuidString }.filter { supported.contains($0) })
    guard !advertised.isEmpty else { return }
    peripherals[peripheral.identifier]=peripheral; advertisedProfiles[peripheral.identifier]=advertised
  }
  func centralManager(_ central:CBCentralManager,didConnect peripheral:CBPeripheral){
    pendingPairResults.removeValue(forKey: peripheral.identifier)?(nil)
    peripheral.delegate=self
    if sessions[peripheral.identifier] != nil { peripheral.discoverServices(supportedServices) }
  }
  func centralManager(_ central:CBCentralManager,didFailToConnect peripheral:CBPeripheral,error:Error?){
    pendingPairResults.removeValue(forKey: peripheral.identifier)?(FlutterError(code:"pairing_failed",message:error?.localizedDescription,details:nil))
    finish(peripheral.identifier,error:FlutterError(code:"gatt_connection_failed",message:error?.localizedDescription,details:nil))
  }
  func centralManager(_ central:CBCentralManager,didDisconnectPeripheral peripheral:CBPeripheral,error:Error?){ if sessions[peripheral.identifier] != nil { finish(peripheral.identifier,error:error == nil ? nil : FlutterError(code:"gatt_disconnected",message:error?.localizedDescription,details:nil)) } }
  func peripheral(_ peripheral:CBPeripheral,didDiscoverServices error:Error?){
    if let error=error { finish(peripheral.identifier,error:FlutterError(code:"service_discovery_failed",message:error.localizedDescription,details:nil));return }
    guard let session=sessions[peripheral.identifier] else{return}
    let services = peripheral.services ?? []
    session.pendingServiceDiscoveries = services.count
    if services.isEmpty { finish(peripheral.identifier,error:nil); return }
    services.forEach{ peripheral.discoverCharacteristics(nil,for:$0) }
  }
  func peripheral(_ peripheral:CBPeripheral,didDiscoverCharacteristicsFor service:CBService,error:Error?){
    guard let session=sessions[peripheral.identifier] else{return}
    if let error=error { finish(peripheral.identifier,error:FlutterError(code:"characteristic_discovery_failed",message:error.localizedDescription,details:nil));return }
    let characteristics=(service.characteristics ?? []).filter {
      supportedCharacteristics.contains($0.uuid.uuidString) &&
        ($0.properties.contains(.read) || $0.properties.contains(.notify) || $0.properties.contains(.indicate))
    }
    for characteristic in characteristics { session.expectedCharacteristics.insert("\(service.uuid.uuidString):\(characteristic.uuid.uuidString)") }
    session.pendingServiceDiscoveries=max(0,session.pendingServiceDiscoveries-1)
    if session.pendingServiceDiscoveries == 0 {
      if session.expectedCharacteristics.isEmpty { finish(peripheral.identifier,error:FlutterError(code:"no_supported_characteristic",message:nil,details:nil)); return }
      for discoveredService in peripheral.services ?? [] {
        for characteristic in discoveredService.characteristics ?? [] where session.expectedCharacteristics.contains("\(discoveredService.uuid.uuidString):\(characteristic.uuid.uuidString)") {
          if characteristic.properties.contains(.notify) || characteristic.properties.contains(.indicate) { peripheral.setNotifyValue(true,for:characteristic) }
          if characteristic.properties.contains(.read) { peripheral.readValue(for:characteristic) }
        }
      }
    }
  }
  func peripheral(_ peripheral:CBPeripheral,didUpdateValueFor characteristic:CBCharacteristic,error:Error?){
    guard let session=sessions[peripheral.identifier] else{return}
    let identity="\(characteristic.service?.uuid.uuidString ?? ""):" + characteristic.uuid.uuidString
    guard session.expectedCharacteristics.contains(identity), !session.completedCharacteristics.contains(identity) else{return}
    session.completedCharacteristics.insert(identity)
    if error == nil, let data=characteristic.value {
      let key="\(peripheral.identifier)-\(characteristic.uuid)-\(data.base64EncodedString())"
      if seenPackets.insert(key).inserted { session.packets.append(["peripheralId":peripheral.identifier.uuidString,"service":characteristic.service?.uuid.uuidString ?? "","characteristic":characteristic.uuid.uuidString,"packet":data.base64EncodedString(),"receivedAt":ISO8601DateFormatter().string(from:Date())]) }
    }
    completeIfReady(peripheral.identifier)
  }
  private func completeIfReady(_ id:UUID){ guard let session=sessions[id],session.pendingServiceDiscoveries==0,session.completedCharacteristics.isSuperset(of:session.expectedCharacteristics) else{return}; finish(id,error:nil) }
  private func finish(_ id:UUID,error:FlutterError?){ guard let session=sessions.removeValue(forKey:id),!session.completed else{return};session.completed=true;session.timeout?.cancel();if let error=error{session.result(error)}else{session.result(session.packets)} }
  private func describe(_ p:CBPeripheral)->[String:Any]{["id":p.identifier.uuidString,"name":p.name ?? "Fitness device","manufacturer":"unknown","firmwareVersion":"unknown","profiles":Array(advertisedProfiles[p.identifier] ?? [])]}
}

File: android/app/src/main/kotlin/com/bilhealth/bodyintelligencelog/BILMedicalBleBridge.kt
package com.bilhealth.bodyintelligencelog

import android.Manifest
import android.bluetooth.*
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.content.Context
import android.content.pm.PackageManager
import android.os.Handler
import android.os.Looper
import androidx.core.content.ContextCompat
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean

class BILMedicalBleBridge(private val context: Context, messenger: BinaryMessenger, private val permissionRequester: ((MethodChannel.Result) -> Unit)? = null) {
  private val channel = MethodChannel(messenger, CHANNEL)
  private val manager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
  private val devices = ConcurrentHashMap<String, BluetoothDevice>()
  private val deviceProfiles = ConcurrentHashMap<String, Set<String>>()
  private val sessions = ConcurrentHashMap<String, BluetoothGatt>()
  private val seen = ConcurrentHashMap.newKeySet<String>()
  private val handler = Handler(Looper.getMainLooper())
  // Production scope: compatible fitness scales, body-composition devices,
  // heart-rate sensors and Pulse Oximeter Service (wellness display only).
  private val services = setOf("181D", "181B", "180D")
  private val measurementCharacteristics = setOf("2A9D", "2A9C", "2A37")

  companion object { const val CHANNEL = "bil.global/medical_ble" }

  init {
    channel.setMethodCallHandler { call, result ->
      val adapter = manager.adapter ?: return@setMethodCallHandler result.error("bluetooth_unavailable", null, null)
      if (!adapter.isEnabled) return@setMethodCallHandler result.error("bluetooth_disabled", null, null)
      if (call.method == "requestPermissions") { permissionRequester?.invoke(result) ?: result.error("permission_request_unavailable", null, null); return@setMethodCallHandler }
      if (!hasPermission()) return@setMethodCallHandler result.error("bluetooth_permission_denied", null, null)
      when (call.method) {
        "discover" -> discover(adapter, call.argument<Int>("timeoutMs") ?: 3000, result)
        "pair" -> pair(call.argument<String>("peripheralId"), result)
        "disconnect" -> disconnect(call.argument<String>("peripheralId"), result)
        "cancel" -> disconnect(call.argument<String>("peripheralId"), result)
        "deviceStatus" -> result.success(mapOf(
          "connected" to sessions.containsKey(call.argument<String>("peripheralId")),
          "batteryPercent" to null,
          "batteryVerified" to false,
        ))
        "forget" -> {
          val id = call.argument<String>("peripheralId")
          id?.let {
            val gatt = sessions.remove(it)
            gatt?.disconnect()
            gatt?.close()
            devices.remove(it)
            deviceProfiles.remove(it)
          }
          result.success(mapOf("forgottenLocally" to true, "systemUnpairRequired" to true))
        }
        "readMeasurements" -> read(call.argument<String>("peripheralId"), result)
        else -> result.notImplemented()
      }
    }
  }

  private fun hasPermission(): Boolean = if (android.os.Build.VERSION.SDK_INT >= 31) {
    ContextCompat.checkSelfPermission(context, Manifest.permission.BLUETOOTH_SCAN) == PackageManager.PERMISSION_GRANTED &&
      ContextCompat.checkSelfPermission(context, Manifest.permission.BLUETOOTH_CONNECT) == PackageManager.PERMISSION_GRANTED
  } else {
    ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED
  }

  private fun discover(adapter: BluetoothAdapter, timeoutMs: Int, result: MethodChannel.Result) {
    val replied = AtomicBoolean(false)
    val callback = object : ScanCallback() {
      override fun onScanResult(type: Int, scan: ScanResult) {
        val advertised = (scan.scanRecord?.serviceUuids ?: emptyList())
          .map { shortUuid(it.uuid.toString()) }
          .filter { services.contains(it) }
          .toSet()
        if (advertised.isNotEmpty()) {
          devices[scan.device.address] = scan.device
          deviceProfiles[scan.device.address] = advertised
        }
      }
      override fun onScanFailed(errorCode: Int) { if (replied.compareAndSet(false, true)) result.error("scan_failed", errorCode.toString(), null) }
    }
    adapter.bluetoothLeScanner.startScan(callback)
    handler.postDelayed({
      adapter.bluetoothLeScanner.stopScan(callback)
      if (replied.compareAndSet(false, true)) result.success(devices.values.map(::describe))
    }, timeoutMs.coerceIn(500, 30000).toLong())
  }

  private fun pair(id: String?, result: MethodChannel.Result) {
    // Standard fitness GATT sensors commonly do not create an Android system
    // bond. The real connection and any platform security prompt happen in
    // read(), so discovery must not fail merely because createBond() is not
    // supported by the peripheral.
    if (id?.let(devices::get) == null) return result.error("device_not_found", null, null)
    result.success(null)
  }

  private fun disconnect(id: String?, result: MethodChannel.Result) {
    id?.let { sessions.remove(it) }?.let { it.disconnect(); it.close() }
    result.success(null)
  }

  private fun read(id: String?, result: MethodChannel.Result) {
    val device = id?.let(devices::get) ?: return result.error("device_not_found", null, null)
    sessions.remove(device.address)?.let { it.disconnect(); it.close() }
    val replied = AtomicBoolean(false)
    val awaitingNotification = AtomicBoolean(false)
    val packets = mutableListOf<Map<String, Any>>()
    val pending = mutableListOf<BluetoothGattCharacteristic>()
    fun finish(error: String? = null) {
      if (!replied.compareAndSet(false, true)) return
      sessions.remove(device.address)?.let { it.disconnect(); it.close() }
      if (error == null) result.success(packets.toList()) else result.error(error, null, null)
    }
    val callback = object : BluetoothGattCallback() {
      override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, state: Int) {
        if (status != BluetoothGatt.GATT_SUCCESS) return finish("gatt_connection_failed")
        if (state == BluetoothProfile.STATE_CONNECTED) gatt.discoverServices()
        else if (state == BluetoothProfile.STATE_DISCONNECTED) finish(if (packets.isEmpty()) "gatt_disconnected" else null)
      }
      override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
        if (status != BluetoothGatt.GATT_SUCCESS) return finish("service_discovery_failed")
        pending += gatt.services
          .filter { services.contains(shortUuid(it.uuid.toString())) }
          .flatMap { service ->
            service.characteristics.filter { characteristic ->
              val supported = measurementCharacteristics.contains(shortUuid(characteristic.uuid.toString()))
              val readable = (characteristic.properties and BluetoothGattCharacteristic.PROPERTY_READ) != 0
              val notifiable = (characteristic.properties and (BluetoothGattCharacteristic.PROPERTY_NOTIFY or BluetoothGattCharacteristic.PROPERTY_INDICATE)) != 0
              supported && (readable || notifiable)
            }
          }
        if (pending.isEmpty()) return finish("no_supported_characteristic")
        readNext(gatt, pending, packets, awaitingNotification, ::finish)
      }
      @Deprecated("Deprecated by Android API; retained for min SDK compatibility")
      override fun onCharacteristicRead(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, status: Int) {
        onPacket(gatt, characteristic, characteristic.value ?: byteArrayOf(), status, packets, pending, awaitingNotification, ::finish)
      }
      override fun onCharacteristicRead(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, value: ByteArray, status: Int) {
        onPacket(gatt, characteristic, value, status, packets, pending, awaitingNotification, ::finish)
      }
      @Deprecated("Deprecated by Android API; retained for min SDK compatibility")
      override fun onCharacteristicChanged(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
        onNotification(gatt, characteristic, characteristic.value ?: byteArrayOf(), packets, ::finish)
      }
      override fun onCharacteristicChanged(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, value: ByteArray) {
        onNotification(gatt, characteristic, value, packets, ::finish)
      }
      override fun onDescriptorWrite(gatt: BluetoothGatt, descriptor: BluetoothGattDescriptor, status: Int) {
        if (status != BluetoothGatt.GATT_SUCCESS) return finish("notification_subscription_failed")
        readNext(gatt, pending, packets, awaitingNotification, ::finish)
      }
    }
    val gatt = device.connectGatt(context, false, callback, BluetoothDevice.TRANSPORT_LE)
    sessions[device.address] = gatt
    handler.postDelayed({ finish("gatt_timeout") }, 20000)
  }

  private fun onPacket(gatt: BluetoothGatt, c: BluetoothGattCharacteristic, value: ByteArray, status: Int, packets: MutableList<Map<String, Any>>, pending: MutableList<BluetoothGattCharacteristic>, awaitingNotification: AtomicBoolean, finish: (String?) -> Unit) {
    if (status == BluetoothGatt.GATT_SUCCESS && value.isNotEmpty()) {
      val key = "${gatt.device.address}:${c.uuid}:${value.contentHashCode()}"
      if (seen.add(key)) packets += mapOf("peripheralId" to gatt.device.address, "service" to c.service.uuid.toString(), "characteristic" to c.uuid.toString(), "packet" to android.util.Base64.encodeToString(value, android.util.Base64.NO_WRAP), "receivedAt" to Instant.now().toString())
    }
    readNext(gatt, pending, packets, awaitingNotification, finish)
  }

  private fun onNotification(gatt: BluetoothGatt, c: BluetoothGattCharacteristic, value: ByteArray, packets: MutableList<Map<String, Any>>, finish: (String?) -> Unit) {
    if (value.isEmpty()) return
    val key = "${gatt.device.address}:${c.uuid}:${value.contentHashCode()}"
    if (seen.add(key)) packets += mapOf("peripheralId" to gatt.device.address, "service" to c.service.uuid.toString(), "characteristic" to c.uuid.toString(), "packet" to android.util.Base64.encodeToString(value, android.util.Base64.NO_WRAP), "receivedAt" to Instant.now().toString())
    finish(null)
  }

  @Suppress("DEPRECATION")
  private fun readNext(gatt: BluetoothGatt, pending: MutableList<BluetoothGattCharacteristic>, packets: MutableList<Map<String, Any>>, awaitingNotification: AtomicBoolean, finish: (String?) -> Unit) {
    if (pending.isEmpty()) {
      if (packets.isNotEmpty()) finish(null)
      else if (!awaitingNotification.get()) finish("no_measurement_received")
      return
    }
    val next = pending.removeAt(0)
    if ((next.properties and BluetoothGattCharacteristic.PROPERTY_READ) != 0) {
      if (!gatt.readCharacteristic(next)) finish("characteristic_read_start_failed")
      return
    }
    val indicate = (next.properties and BluetoothGattCharacteristic.PROPERTY_INDICATE) != 0
    val descriptor = next.getDescriptor(java.util.UUID.fromString("00002902-0000-1000-8000-00805f9b34fb"))
    if (descriptor == null || !gatt.setCharacteristicNotification(next, true)) {
      finish("notification_subscription_unavailable")
      return
    }
    awaitingNotification.set(true)
    descriptor.value = if (indicate) BluetoothGattDescriptor.ENABLE_INDICATION_VALUE else BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
    if (!gatt.writeDescriptor(descriptor)) finish("notification_subscription_failed")
  }

  private fun shortUuid(value: String) = value.substring(4, 8).uppercase()
  private fun describe(device: BluetoothDevice): Map<String, Any> = mapOf(
    "id" to device.address,
    "name" to (device.name ?: "Fitness device"),
    "manufacturer" to "unknown",
    "firmwareVersion" to "unknown",
    "profiles" to (deviceProfiles[device.address] ?: emptySet<String>()).toList(),
  )
}
